src/protean/core/entity.py

```diff
@@ -89,6 +89,164 @@ def has_auto_field(self):
                     _, field_obj in self.declared_fields.items()])
 
 
+class FilterSet:
+    """A chainable class to gather a bunch of Filters before execution
+
+    Internally, a Filterset can be constructed, filtered, sliced, and generally passed around
+    without actually fetching data. No data fetch actually occurs until you do something
+    to evaluate the queryset.
+    """
+
+    def __init__(self, entity_cls_name: str, page: int = 1, per_page: int = 10,
+                 order_by: set = None, excludes_: dict = None, **filters):
+        """Initialize either with empty filters (when invoked on an Entity)
+            or carry forward filters when chained
+        """
+
+        self._entity_cls_name = entity_cls_name
+        self._page = page or 1
+        self._per_page = per_page or 10
+
+        # `order_by` could be empty, or a string or a set.
+        #   Intialize empty set if `order_by` is None
+        #   Convert string to set if `order_by` is a String
+        #   Safe-cast set to a set if `order_by` is already a set
+        if order_by:
+            self._order_by = set([order_by]) if isinstance(order_by, str) else set(order_by)
+        else:
+            self._order_by = set()
+        self._excludes = excludes_ or {}
+        self._filters = filters
+
+    def filter(self, **filters):
+        """Merge new filter list with existing filters"""
+        self._filters.update(filters)
+
+        return self
+
+    def exclude(self, **excludes):
+        """Merge new exclude list with existing excludes dictionary"""
+        self._excludes.update(excludes)
+
+        return self
+
+    def page(self, page: int):
+        """Update page setting for filter set"""
+        self._page = page
+
+        return self
+
+    def per_page(self, per_page: int):
+        """Update per_page setting for filter set"""
+        self._per_page = per_page
+
+        return self
+
+    def order_by(self, order_by: Union[set, str]):
+        """Update page setting for filter set"""
+        if isinstance(order_by, str):
+            order_by = {order_by}
+
+        self._order_by = self._order_by.union(order_by)
+
+        return self
+
+    def _retrieve_model(self):
+        """Retrieve model details associated with this Entity"""
+        from protean.core.repository import repo_factory  # FIXME Move to a better placement
+
+        # Fetch Model class and connected-adapter from Repository Factory
+        model_cls = repo_factory.get_model(self._entity_cls_name)
+        adapter = getattr(repo_factory, self._entity_cls_name)
+
+        return (model_cls, adapter)
+
+    def values(self):
```

Comment:
`execute` or `evaluate` are verbs, which I want this function to denote the results. But this function will be rarely invoked directly. We can make it private. Accessing the query in a certain way, like calling `list()` of it, or slicing it, or trying to iterate on it, will trigger this function automatically.

Reply:
Yes that will be better @subhashb

Reply:
The real reason `values()` is dangerous is because `dict` has an implementation of `values()` and it's not clear which method is being invoked on the results. Renamed the method to `all()` following implementations in Django, SQLAlchemy, Rails etc.

Reply:
Changed the method name to `all()` and pushed as part of https://github.com/proteanhq/protean/commit/ae5339018972f0201be2f155d9b268590579d7f3